### PR TITLE
Bug 1864070 - Add secret setting to enable the debug drawer

### PR DESCRIPTION
**Pull request by MozillaNoah**
After some iteration of exploring the use of the legacy Preferences API + AppState, I've opted to use DataStore here instead. With this approach, we'll have an API that's easy to consume in Compose while also not introducing a huge hack into SecretSettingsFragment in order to handle our use case of an observable preference.

[DataStore is currently being leveraged for Pocket](https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/datastore/DataStores.kt), so there is no library addition here.

Hopefully, this will serve as a potential migration path for the rest of our settings. 

In order to get the secret settings into the debug drawer, I'll likely need to convert all of them to Compose + DataStore.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
4. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
5. Click on `View task in Taskcluster` in the new `DETAILS` section.
6. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1864070

### CODE REVIEW

File: fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt [hunk 1]
```diff
@@ -86,6 +87,18 @@ class SecretSettingsFragment : PreferenceFragmentCompat() {
...
 
+        requirePreference<SwitchPreference>(R.string.pref_key_enable_debug_drawer).apply {
+            isVisible = Config.channel.isNightlyOrDebug
+            isChecked = context.settings().enableDebugDrawer
+            onPreferenceChangeListener =
+                Preference.OnPreferenceChangeListener { _, newValue ->
+                    context.settings().enableDebugDrawer = newValue as Boolean
+                    context.components.appStore.dispatch(AppAction.UpdateDebugDrawerEnabled(newValue))
+
+                    true
```
Comment (MatthewTighe) [inline fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:98]: IMO if we are going to manage this state in a Store, then a middleware should be handling the `Settings` side-effects

**Review by MatthewTighe — COMMENTED** [review]
> 1. Does it make sense to have debugging state inside of AppState? Do we already have precedence for this?

I'm not seeing that we do have precedence for it. At first glance, my intuition is telling me that a separate Store for the Debug Drawer would be more appropriate. I think this could potentially benefit from prototyping though. The open questions in my mind are things like: what kind of other state would live in it? How would that state be used? 

A few cases that seem like they could provide interesting learnings, which coincidentally include your existing plans:
1. the tabs management you're already planning. This will naturally need to have an effect on the browser store, and experimenting with the communication between the two components seems informative
2. Overwriting secret settings. This begs the question of where these preferences live - now they're accessed directly, but do we want to move towards a world where our Stores provide the interface to our persisted properties (my answer would be yes, but then we have more work to do)?
3. Nimbus features. This demonstrates a similar problem to `2`, in that we'll need to explore how we manipulate the existing source of truth with the debug drawer.

Tangentially, creating a separate Store may make it easier to disable when not in use. And if we ever move towards Scoped Stores, it may be easier to combine disparate stores than try to split out a monolith into the appropriate scopes (this is very hypothetical opinion).

> 2. If we do want to have a debugging state inside of AppState, what would be the best way to manage this? Right now, we might only need a boolean for the debug drawer, but we could easily foresee a future where more debugging state information is kept at the app level for whatever purpose. In that case, should we instead have a substate, a DebugState if you will, similar to ShoppingState and WallpaperState?

I think if we wanted to include it in AppStore for now, than it should absolutely be encapsulated into a distinct `DebugState`. In general I think the AppStore needs to broken apart a bit more, either into more distinct states (Pocket) or even more Stores (HomeFragment).

**Comment by MozillaNoah** [comment]
After discussing with @gabrielluong, I've just pushed a new version of the draft which utilizes `DataStore` instead of shared preferences. An `AppAction` triggers the data store write. 

Depending how this iteration goes. I may need to port the rest of the secret settings to `DataStore` in order to create the Composified UI in the debug drawer.

File: fenix/app/src/main/java/org/mozilla/fenix/components/appstate/AppStoreReducer.kt [hunk 2]
```diff
@@ -234,6 +234,8 @@ internal object AppStoreReducer {
         )
 
         is AppAction.ShoppingAction -> ShoppingStateReducer.reduce(state, action)
+
+        is AppAction.UpdateDebugDrawerEnabled -> state.copy() // consumed by preferences middleware
```
Comment (MatthewTighe) [inline fenix/app/src/main/java/org/mozilla/fenix/components/appstate/AppStoreReducer.kt:238]:
```suggestion
        is AppAction.UpdateDebugDrawerEnabled -> state // consumed by preferences middleware
```

File: fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt [hunk 3]
```diff
@@ -28,6 +34,11 @@ class SecretSettingsFragment : PreferenceFragmentCompat() {
     }
 
     override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
+        var debugDrawerEnabled = false
+        lifecycleScope.launch {
+            debugDrawerEnabled = context?.debugSettings?.debugDrawerEnabledFlow?.first() ?: false
```
Comment (MatthewTighe) [inline fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:39]: Do we want to collect the flow rather than just take the value out of it so that we get reactivity?

File: fenix/app/src/main/java/org/mozilla/fenix/components/appstate/PreferencesMiddleware.kt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    private val debugSettingsDataStore: DataStore<Preferences>,
+    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO),
+) : Middleware<AppState, AppAction> {
+
+    override fun invoke(
+        context: MiddlewareContext<AppState, AppAction>,
+        next: (AppAction) -> Unit,
+        action: AppAction,
+    ) {
+        handleAction(action)
```
Comment (MatthewTighe) [inline fenix/app/src/main/java/org/mozilla/fenix/components/appstate/PreferencesMiddleware.kt:32]: Since we can probably assume that the actions here are all disk writes, it might help convey some semantic value to do `scope.launch { handleAction(action) }` and make `handleAction` suspending. It would send a message to future readers that `handleAction` doesn't block `next`.

File: fenix/app/src/main/java/org/mozilla/fenix/components/appstate/AppAction.kt [hunk 5]
```diff
@@ -262,4 +262,11 @@ sealed class AppAction : Action {
             val expanded: Boolean,
         ) : ShoppingAction()
     }
+
+    /**
+     * Action used to update the visibility of the Debug Drawer trigger.
+     *
+     * @property enabled Whether the debug drawer has been enabled as part of this Action.
+     */
+    data class UpdateDebugDrawerEnabled(val enabled: Boolean) : AppAction()
```
Comment (MatthewTighe) [inline fenix/app/src/main/java/org/mozilla/fenix/components/appstate/AppAction.kt:271]: If we aren't keeping the enabled state in the Store, I'm not sure if we need to use an action/middleware to control the side-effects

**Review by MatthewTighe — COMMENTED** [review]
Reflecting more on my earlier comment:

> At first glance, my intuition is telling me that a separate Store for the Debug Drawer would be more appropriate.

I think that I may have been wrong there. I'm not sure it makes sense to keep a 3rd global store around, since there seems to be a high likelihood that we will want to pass it into various components that we're already passing the `AppStore`.

So I think two options could be:
1. `DebugState` as part of `AppState`. I like that this is in keeping with the rest of our architecture, and it could even give us a bit of a playground to test future patterns.
2. Directly accessing debug state through the datastore helpers. I like that this separates code that governs "internal" behavior and user-facing behavior. 

I am maybe a little more inclined towards the first option at this point, but I'll leave it up to you what you want to try first. There could be other options too, I'm not sure that's exhaustive. I think it could be good to get a second opinion here, since this is having a bit more of an impact on discussions of architecture.

File: fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt (hunk 3 above)
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:39]: I only need the one-time read, since the UI will update correctly with the user's input

File: fenix/app/src/main/java/org/mozilla/fenix/components/appstate/AppAction.kt (hunk 5 above)
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/components/appstate/AppAction.kt:271]: That's fair. Since this doesn't need any other side effects, the Action/Middleware is a little heavy handed.

**Comment by MozillaNoah** [comment]
> So I think two options could be:
> 
>     1. `DebugState` as part of `AppState`. I like that this is in keeping with the rest of our architecture, and it could even give us a bit of a playground to test future patterns.
> 
>     2. Directly accessing debug state through the datastore helpers. I like that this separates code that governs "internal" behavior and user-facing behavior.
> 
> 
> I am maybe a little more inclined towards the first option at this point, but I'll leave it up to you what you want to try first. There could be other options too, I'm not sure that's exhaustive. I think it could be good to get a second opinion here, since this is having a bit more of an impact on discussions of architecture.

I think at this point it's a matter of trade offs

1. `DebugState` as part of `AppState`.
    -  Keeps the state global
    - Has the pattern we already know how to implement and consume
    - The state is really living in two places simultaneously (shared preferences & app state)
2. Using `DataStore` as the proof of concept observable persistent preferences.
    - This is a new pattern and preference architecture ([it is used for Pocket though](https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/datastore/DataStores.kt#14))
    - Non-zero learning curve
    - Easily consumable in Compose

Either way, there will be some degree of clunkiness because we're dealing with the legacy XML prefs and wanting to live in a modern, observable world. As it stands right now, I'll likely need and want to convert Secret Settings as a whole into DataStore, since the preferences API so tightly coupled to XML and the shared preferences disk writing, preventing me to do a simple "migration" to Compose from a preferences XML file and PreferenceFragment.

File: fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt (hunk 3 above)
Comment (MatthewTighe) [reply fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:39]: Is the read guaranteed to complete before we finish setting up the UI initially? Since we're launching a coroutine to collect the read in, it seems like we could reach the end up setup before `debugDrawerEnabled` has been updated the result

Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:39]: I don't believe so, but we have a couple options here:
- Use `runBlocking` or our `runBlockingIncrement` for this singular read
- Pre-load the data into memory ([doc](https://developer.android.com/topic/libraries/architecture/datastore#synchronous))

For option 2, we just fire off pre-read of the store inside of onCreate, essentially.

File: fenix/app/src/main/java/org/mozilla/fenix/datastore/DebugDataStores.kt
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/datastore/DebugDataStores.kt:1]: What do you think about calling this`DebugDrawerDataStore`? `DebugDataStores` sounds a little generic and not sure if the pluralizing is needed?

File: fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt (hunk 3 above)
Comment (t-p-white) [reply fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:39]: Seems like option 2 would be preferable based on the doc? 

I haven't tested this but throwing out there, could you wrap the preference in a `lifecycleScope.launch` such as:
```
 lifecycleScope.launch {
            val debugDrawerEnabled = context?.debugSettings?.isDebugDrawerEnabled?.first() ?: false

            requirePreference<SwitchPreference>(R.string.pref_key_enable_debug_drawer).apply {
                isVisible = Config.channel.isNightlyOrDebug
                isChecked = debugDrawerEnabled
                onPreferenceChangeListener =
                    Preference.OnPreferenceChangeListener { _, newValue ->
                        lifecycleScope.launch {
                            context.debugSettings.updateDebugDrawerEnabled(enabled = newValue as Boolean)
                        }
                        true
                    }
            }
        }
``` 
or will this have an issue drawing async?

File: fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt [hunk 6]
```diff
@@ -86,6 +97,18 @@ class SecretSettingsFragment : PreferenceFragmentCompat() {
             }
         }
```
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:99]: Might be nice to include a small comment similar to the block below? e.g.  "We only want to enable the debug drawer for Nightly or Debug builds because..."

**Review by t-p-white — COMMENTED** [review]
Looking good Noah! Just a few comments from me

File: fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt (hunk 3 above)
Comment (t-p-white) [reply fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:39]: nit: I also wonder whether the `?.first() ?: false` should be move directly to `debugDrawerEnabledFlow`

File: fenix/app/src/main/java/org/mozilla/fenix/datastore/DebugDataStores.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/datastore/DebugDataStores.kt:1]: The pluralization was on accident, so thanks for catching that.

I initially made it just `debug` as an intentional future catch all, but having the specifically-scoped naming makes more sense looking at it.

File: fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt (hunk 3 above)
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:39]: I'll give the preference wrap a try.

I'll move it into the data store file and use `false` as a default static value

File: fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt [hunk 7]
```diff
@@ -92,6 +98,23 @@ class SecretSettingsFragment : PreferenceFragmentCompat() {
             onPreferenceChangeListener = SharedPreferenceUpdater()
         }
 
+        lifecycleScope.launch {
+            val debugDataStore = requireContext().debugSettings
+
+            // During initial development, this will only be available in Nightly or Debug builds.
+            requirePreference<SwitchPreference>(R.string.pref_key_enable_debug_drawer).apply {
+                isVisible = Config.channel.isNightlyOrDebug
+                isChecked = debugDataStore.debugDrawerEnabled.first()
```
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:107]: nit: IMO I think moving the `first()` call directly to `debugDrawerEnabled` would keep the responsibility in a single place instead of the caller having to make the call on a `flow` as we only want the boolean value and the flow won't update after the first call? e.g.
```kotlin
suspend fun DataStore<Preferences>.debugDrawerEnabled(): Boolean = data.map { preferences ->
        preferences[DEBUG_DRAWER_ENABLED] ?: false
    }.first()
```
Not a blocker though

File: fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt [hunk 8]
```diff
@@ -92,6 +98,23 @@ class SecretSettingsFragment : PreferenceFragmentCompat() {
             onPreferenceChangeListener = SharedPreferenceUpdater()
         }
 
+        lifecycleScope.launch {
+            val debugDataStore = requireContext().debugSettings
```
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:102]: nit: this could maybe be inlined if the suggestion below is used

**Review by t-p-white — COMMENTED** [review]
>This is a new pattern and preference architecture ([it is used for Pocket though](https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/datastore/DataStores.kt#14))

Do we know why it was used for Pocket?

File: fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt (hunk 7 above)
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:107]: We'll need the observe the flow as part of https://github.com/mozilla-mobile/firefox-android/pull/4612/commits/32fc424e1482913e97c21c478ae88062b9bed65b. 

Should we keep the `single` in that case, or maybe instead just expose a non-observable API?

Comment (t-p-white) [reply fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:107]: Ah okay, so just checking my understanding:
* Prefs screen needs the `Boolean` value only as this retrieves the value only when created
* The overlay needs to observe the `Flow` to update its enabled state as this is always present
?

**Comment by t-p-white** [comment]
@MozillaNoah after having a bit more of a read through of the `DataStore` docs, if we are going for:

>     2. Using `DataStore` as the proof of concept observable persistent preferences.
>        
>        * This is a new pattern and preference architecture ([it is used for Pocket though](https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/datastore/DataStores.kt#14))
>        * Non-zero learning curve
>        * Easily consumable in Compose

Should we also establish a pattern where-by we use a top level 'grouping' DataStore such as a 'secret settings preferences' DataStore in `DataStores` e.g.:

```kotlin
/**
 * Class that handles saving and retrieving secret settings preferences
 */

private const val SECRET_SETTING_PREFERENCES_NAME = "secret_settings_preferences"

// Create a DataStore instance using the preferencesDataStore delegate, with the Context as
// receiver.
val Context.dataStore: DataStore<Preferences> by preferencesDataStore(
    name = SECRET_SETTING_PREFERENCES_NAME
)
```
For reference, I was using the guide and code from this [CodeLab](https://github.com/google-developer-training/android-basics-kotlin-words-app/blob/datastore/app/src/main/java/com/example/wordsapp/data/SettingsDataStore.kt).

**Comment by MatthewTighe** [comment]
Just wanting share some thoughts about a couple things in the thread above, but I don't think we need to spend too long analyzing this as we iterate on it. Since this isn't user-facing or critical path at all, it should provide an easy area for more rapid adjustment than elsewhere in the app.

>1. DebugState as part of AppState.
> - Keeps the state global
> - Has the pattern we already know how to implement and consume
> - The state is really living in two places simultaneously (shared preferences & app state)
> - Keeps us stuck with the monolithic Preferences API

Specifically:

> The state is really living in two places simultaneously (shared preferences & app state)

This is true of any state whose real source of truth is behind an async API. It's also true for architectures that aren't ours, like MVVM. The issue is that we often want to update the local state before we can fetch the latest from async (whether disk or network), but this runs risk of divergence. 

The Redux pattern tends to encourage more eagerly updating local state (through the reducer), but I don't think it's _required_. We could instead:
1. Dispatch an action like `DebugDrawerPreferenceChanged` 
2. Write the update to disk from a middleware
3. Ignore the action in the reducer
4. Observe the datastore from some kind of binding which then updates the store with like `DebugDrawerPreferenceRead`
5. the reducer uses the result of the datastore through the second action

I think, generally speaking, we just trust that the async operations succeed and that the remote and local sources of truth stay in sync. There's probably lots to explore here in terms of out-of-sync error handling, caching mechanisms, etc.

I don't think we need to do any of this for this use case now (or ever if we don't want to), but I do think  in general that `DataStore` state would be best served by living in a Store.

**Review by MatthewTighe — APPROVED** [review]
This LGTM as a first iteration based on our current goals. As I mention in my comment above, this code should be easier to change down the line than other code in the app.

I do want to also mention that as we switch to DataStore we may want to create abstract super types for testing. Something like:

```
interface DebugSettingsCache {
    fun debugDrawerEnabledFlow(): Flow<Boolean>
    fun setDebugDrawerEnabled(isEnabled: Boolean)
}

class DebugSettingsDataStore(
    context: Context,
    private val writeScope: CoroutineScope = CoroutineScope(Dispatchers.IO)
) : DebugSettingsCache {
    private val debugCache = context.debugSettings
    override fun debugDrawerEnabledFlow(): Flow<Boolean> =
        debugCache.data.map { preferences ->
            preferences[DEBUG_DRAWER_ENABLED] ?: false
        }

    override fun setDebugDrawerEnabled(isEnabled: Boolean) {
        writeScope.launch {
            debugCache.edit { preferences ->
                preferences[DEBUG_DRAWER_ENABLED] = isEnabled
            }
        }
    }
}
```

This would just make swapping out a fake in tests much easier. Otherwise, I think we'd have to like mock the return value of the extension function you've defined on `Context`? 

Again, probably not necessary for this use case but we should be careful that this isn't used as a definitive precedent down the line.

File: fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt (hunk 7 above)
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/settings/SecretSettingsFragment.kt:107]: That's correct!

**Comment by MozillaNoah** [comment]
> @MozillaNoah after having a bit more of a read through of the `DataStore` docs, if we are going for:
> 
> > ```
> > 2. Using `DataStore` as the proof of concept observable persistent preferences.
> >    
> >    * This is a new pattern and preference architecture ([it is used for Pocket though](https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/datastore/DataStores.kt#14))
> >    * Non-zero learning curve
> >    * Easily consumable in Compose
> > ```
> 
> Should we also establish a pattern where-by we use a top level 'grouping' DataStore such as a 'secret settings preferences' DataStore in `DataStores` e.g.:
>
> For reference, I was using the guide and code from this [CodeLab](https://github.com/google-developer-training/android-basics-kotlin-words-app/blob/datastore/app/src/main/java/com/example/wordsapp/data/SettingsDataStore.kt).

That's certainly an option! I'm going to investigate Secret Settings in general as part of [Bug 1864712](https://bugzilla.mozilla.org/show_bug.cgi?id=1864712), but I currently foresee two ways to group our preferences that will be maintained via `DataStore`.

1. Have a data store to mirror the UI of our settings flows (e.g. data store for tabs, data store for search, etc.)
2. Have a data store scoped to the use case/purpose (e.g. debugging, everything related to tabs, 

Basically, do we couple the data stores to their code adjacency or their product adjacency? We don't have to solve this now, but this work and [Bug 1864712](https://bugzilla.mozilla.org/show_bug.cgi?id=1864712) will undoubtedly inform us how to proceed with porting the rest of `Settings.kt` and creating new preferences going forward.

**Comment by MozillaNoah** [comment]
> This LGTM as a first iteration based on our current goals. As I mention in my comment above, this code should be easier to change down the line than other code in the app.
> 
> I do want to also mention that as we switch to DataStore we may want to create abstract super types for testing. Something like:
> 
> ```
> interface DebugSettingsCache {
>        ...
> }
> 
> class DebugSettingsDataStore(
>     context: Context,
>     private val writeScope: CoroutineScope = CoroutineScope(Dispatchers.IO)
> ) : DebugSettingsCache {
>     ...
> }
> ```
> 
> This would just make swapping out a fake in tests much easier. Otherwise, I think we'd have to like mock the return value of the extension function you've defined on `Context`?
> 
> Again, probably not necessary for this use case but we should be careful that this isn't used as a definitive precedent down the line.

I love this idea! I'll write a ticket to add this abstraction. It will certainly make tests easier when the redux layer is set up as part of [Bug 1864073](https://bugzilla.mozilla.org/show_bug.cgi?id=1864073) and then we can test the data store in isolation!

**Comment by MozillaNoah** [comment]
Hey @t-p-white, I'm going to land this today, but let's continue the discussion of DataStores and preferences as this feature evolves!
